src/Database/LSMTree/Internal/BlobRef.hs

```diff
@@ -83,45 +92,45 @@ newtype WeakBlobRefInvalid = WeakBlobRefInvalid Int
 withWeakBlobRef ::
      (MonadMask m, PrimMonad m)
   => WeakBlobRef m h
-  -> (BlobRef m h -> m a)
+  -> (RawBlobRef m h -> m a)
   -> m a
 withWeakBlobRef wref = bracket (deRefWeakBlobRef wref) removeReference
 
 {-# SPECIALISE withWeakBlobRefs ::
      V.Vector (WeakBlobRef IO h)
-  -> (V.Vector (BlobRef IO h) -> IO a)
+  -> (V.Vector (RawBlobRef IO h) -> IO a)
   -> IO a #-}
 -- | The same as 'withWeakBlobRef' but for many references in one go.
 --
 withWeakBlobRefs ::
      (MonadMask m, PrimMonad m)
   => V.Vector (WeakBlobRef m h)
-  -> (V.Vector (BlobRef m h) -> m a)
+  -> (V.Vector (RawBlobRef m h) -> m a)
   -> m a
 withWeakBlobRefs wrefs = bracket (deRefWeakBlobRefs wrefs) removeReferences
 
 {-# SPECIALISE deRefWeakBlobRef ::
      WeakBlobRef IO h
-  -> IO (BlobRef IO h) #-}
+  -> IO (RawBlobRef IO h) #-}
 deRefWeakBlobRef ::
      (MonadThrow m, PrimMonad m)
   => WeakBlobRef m h
-  -> m (BlobRef m h)
+  -> m (RawBlobRef m h)
 deRefWeakBlobRef (WeakBlobRef ref) = do
     ok <- RC.upgradeWeakReference (blobRefCount ref)
     when (not ok) $ throwIO (WeakBlobRefInvalid 0)
     pure ref
```

Comment:
The fact that this returns a `RawBlobRef` seems to be at odds with the haddocks on `RawBlobRef` and `WeakBlobRef`. Those haddocks mention that dereferencing a `WeakBlobRef` should yield a strong blob reference, though it returns a `RawBlobRef`, of which the haddocks mentions it is is not a reference counted object that takes ownership of the `BlobFile`. The implementation of this function *does* use a `RawBlobRef` to take ownership of the `BlobFile` by upgrading the weak reference count. So it seems a `RawBlobRef` serves as both a raw and strong reference. Maybe we should introduce a `StrongBlobRef` newtype/datatype that wraps a `RawBlobRef`?

Reply:
This is an intermediate state. By the end of the patch series `deRefWeakBlobRef` returns a strong reference. At this point in the series, `BlobRef` is still doing dual work as both raw and strong without a type distinction.